### PR TITLE
Add npm/package.json pass

### PR DESCRIPTION
**Pull request by justinfagnani**

### CODE REVIEW

File: cleanup-passes/npm.ts [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+  if (version.startsWith('polymerelements/')) {
+    return `@polymer/${name}`;
+  }
+}
+
+function getPackageVersion(name: string) {
+  if (name === 'polymer') {
+    return '1.2.5-npm-test.1';
+  } else {
+    return '0.0.1';
```
Comment (rictic) [inline cleanup-passes/npm.ts]: :shipit:

File: cleanup-passes/npm.ts [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+function getNpmName(name: string, version: string) {
+  version = version.toLowerCase();
+  if (dependencyMap[name]) {
+    return dependencyMap[name];
+  }
+  if (version.startsWith('polymerelements/')) {
+    return `@polymer/${name}`;
+  }
+}
```
Comment (rictic) [inline cleanup-passes/npm.ts]: consider moving throw here?

File: cleanup-passes/npm.ts [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    return '1.2.5-npm-test.1';
+  } else if (version.startsWith('polymerelements/')) {
+    return '^0.0.1';
+  }
+  return version.substring(version.indexOf('#') + 1);
+}
+
+/**
+ * Cleans up a number of common bower problems, like no "main" attribute,
+ * "main" being an array rather than a string, etc.
```
Comment (rictic) [inline cleanup-passes/npm.ts]: docs

:page_facing_up: :page_facing_up:  :spaghetti:

Comment (justinfagnani) [inline cleanup-passes/npm.ts]: done

File: cleanup-passes/npm.ts (hunk 2 above)
Comment (justinfagnani) [inline cleanup-passes/npm.ts]: done

File: cleanup-passes/npm.ts (hunk 1 above)
Comment (justinfagnani) [inline cleanup-passes/npm.ts]: :ok:

File: cleanup-passes/npm.ts [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+  let bowerConfig: any = readJson('bower.json');
+  let npmConfig: any = readJson('package.json');
+
+  if (!bowerConfig) {
+    throw new Error('bower.json not found');
+  }
+
+  if (bowerConfig['name'] !== 'polymer' && npmConfig['name']) {
+    console.warn(`\nunexpected existing package.json for ${element.repo}`);
```
Comment (rictic) [inline cleanup-passes/npm.ts]: can remove

File: cleanup-passes/npm.ts [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    // throw new Error(`unexpected existing package.json for ${element.repo}`);
+  }
+
+  const copyProperty = (name: string, bowerName?: string) =>
+      npmConfig[name] = bowerConfig[bowerName || name] || npmConfig[name];
```
Comment (rictic) [inline cleanup-passes/npm.ts]: mention bower config is source of truth in docs

File: cleanup-passes/npm.ts [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+  // https://docs.npmjs.com/files/package.json#license
+  npmConfig['license'] = "BSD-3-Clause";
+
+  let npmDependencies = npmConfig['dependencies'] =
+      npmConfig['dependencies'] || {};
+  let bowerDependencies = bowerConfig['dependencies'] || {};
+  let npmDevDependencies = npmConfig['devDependencies'] =
+      npmConfig['devDependencies'] || {};
+  let bowerDevDependencies = bowerConfig['devDependencies'] =
```
Comment (rictic) [inline cleanup-passes/npm.ts]: don't update bowerConfig here

File: cleanup-passes/npm.ts [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+      bowerConfig['devDependencies'] || {};
+
+  npmDevDependencies['web-component-tester'] = '^4.0.0';
+
+  for (let bowerDep in bowerDependencies) {
+    let bowerVersion = bowerDependencies[bowerDep];
+    let npmDep = getNpmName(bowerDep, bowerVersion);
+    if (!npmDep) {
+      // console.warn(`*** no npm name mapping found for ${bowerDep}:${bowerVersion} in ${bowerConfig['name']}`);
```
Comment (rictic) [inline cleanup-passes/npm.ts]: remove

File: cleanup-passes/npm.ts [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+      throw new Error(`no npm name mapping found for ${bowerDep}:${bowerVersion} in ${bowerConfig['name']}`);
+    }
+    npmDependencies[npmDep] = getDependencyVersion(bowerDep, bowerVersion);
+  }
+
+  for (let bowerDep in bowerDevDependencies) {
+    let bowerVersion = bowerDevDependencies[bowerDep];
+    let npmDep = getNpmName(bowerDep, bowerVersion);
+    if (!npmDep) {
+      // console.warn(`*** no npm name mapping found for ${bowerDep}:${bowerVersion} in ${bowerConfig['name']}`);
```
Comment (rictic) [inline cleanup-passes/npm.ts]: remove

File: cleanup-passes/npm.ts (hunk 6 above)
Comment (justinfagnani) [inline cleanup-passes/npm.ts]: done

File: cleanup-passes/npm.ts [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+      throw new Error(`no npm name mapping found for ${bowerDep}:${bowerVersion} in ${bowerConfig['name']}`);
+    }
+    bowerDevDependencies[npmDep] = getDependencyVersion(bowerDep, bowerVersion);
+  }
+
+  console.log(`\npackage.json for ${element.repo}`, JSON.stringify(npmConfig, null, 2));
```
Comment (rictic) [inline cleanup-passes/npm.ts]: remove

File: cleanup-passes/npm.ts (hunk 7 above)
Comment (justinfagnani) [inline cleanup-passes/npm.ts]: done

File: cleanup-passes/npm.ts (hunk 8 above)
Comment (justinfagnani) [inline cleanup-passes/npm.ts]: done

File: cleanup-passes/npm.ts [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+
+  writeJson('package.json', npmConfig);
+
+  // element.needsReview = true;
```
Comment (rictic) [inline cleanup-passes/npm.ts]: remove

File: cleanup-passes/npm.ts [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+  await makeCommit(element, ['package.json'], 'Update package.json');
+}
+
+register({
+  name: 'npm',
```
Comment (rictic) [inline cleanup-passes/npm.ts]: recommend renaming to something like package json generation?

File: tedium.ts [hunk 12]
```diff
@@ -499,7 +499,7 @@ async function _main(elements: ElementRepo[]) {
     'repos/style-guide',
     'repos/test-all',
     'repos/ContributionGuide',
-    'repos/polymer',
+    // 'repos/polymer',
```
Comment (rictic) [inline tedium.ts]: :fire: :skull:  :warning:

File: cleanup-passes/npm.ts (hunk 4 above)
Comment (justinfagnani) [inline cleanup-passes/npm.ts]: done

File: cleanup-passes/npm.ts (hunk 5 above)
Comment (justinfagnani) [inline cleanup-passes/npm.ts]: done

Comment (justinfagnani) [inline cleanup-passes/npm.ts]: done

File: cleanup-passes/npm.ts (hunk 9 above)
Comment (justinfagnani) [inline cleanup-passes/npm.ts]: done

File: cleanup-passes/npm.ts (hunk 10 above)
Comment (justinfagnani) [inline cleanup-passes/npm.ts]: done

File: cleanup-passes/npm.ts (hunk 11 above)
Comment (justinfagnani) [inline cleanup-passes/npm.ts]: done

File: tedium.ts (hunk 12 above)
Comment (justinfagnani) [inline tedium.ts]: done, though paper-listbox appears to work now.

**Comment by justinfagnani** [comment]
PTAL

File: cleanup-passes/package-json.ts [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+    return '^' + polymerVersion;
+  }
+  // I accidentally published iron-ajax with a dep on promise-polyfill 1.0.0
+  if (version.startsWith('polymerlabs/promise-polyfill')) {
+    return '^' + promisePolyfillVersion;
+  }
+  if ((version.startsWith('polymerelements/') ||
+      version.startsWith('polymerlabs/') ||
+      version.startsWith('polymer/')) &&
+      !dependencyMap[name]) {
```
Comment (rictic) [inline cleanup-passes/package-json.ts:81]: When do we want to use the v0.0.3 elementVersion with something which doesn't begin with polymerelements/ polymerlabs/ or polymer/?

**Comment by rictic** [comment]
This looks good, what's the plan for transitioning from the alpha v0.0.x versions to mirroring the bower ones?

**Comment by justinfagnani** [comment]
I think after the next Polymer core release, so that I can set that dependency to a non-pre-release version.

File: cleanup-passes/package-json.ts (hunk 13 above)
Comment (justinfagnani) [inline cleanup-passes/package-json.ts:81]: Never. This logic should be saying that if the version starts with polymerelements/ polymerlabs/ or polymer/, use 0.0.3, unless the package is in the dependency map. I think the dependencyMap check is from an older version of this logic, and unnecessary now.

Comment (rictic) [inline cleanup-passes/package-json.ts:81]: Recommend trying without the dependencyMap check

**Comment by rictic** [comment]
Mind documenting when we expect to remove all of the versioning special casing in the code?

Other than that and the above comment, LGTM :+1:

**Pull request merged**